staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go

```diff
@@ -59,3 +64,82 @@ type InTreePlugin interface {
 	// RepairVolumeHandle generates a correct volume handle based on node ID information.
 	RepairVolumeHandle(volumeHandle, nodeID string) (string, error)
 }
+
+// getTopology returns the current topology zones with the given key.
+func getTopology(pv *v1.PersistentVolume, key string) []string {
+	if pv.Spec.NodeAffinity == nil ||
+		pv.Spec.NodeAffinity.Required == nil ||
+		len(pv.Spec.NodeAffinity.Required.NodeSelectorTerms) < 1 {
+		return nil
+	}
+	for i := range pv.Spec.NodeAffinity.Required.NodeSelectorTerms {
+		for _, r := range pv.Spec.NodeAffinity.Required.NodeSelectorTerms[i].MatchExpressions {
+			if r.Key == key {
+				return r.Values
+			}
+		}
+	}
+	return nil
+}
+
+// recordTopology writes the topology to the given PV.
+// If topology already exists with the given key, assume the content is already accurate.
+func recordTopology(pv *v1.PersistentVolume, key string, zones []string) error {
+	// Make sure there are no duplicate or empty strings
+	filteredZones := sets.String{}
+	for i := range zones {
+		zone := strings.TrimSpace(zones[i])
+		if len(zone) > 0 {
+			filteredZones.Insert(zone)
+		}
+	}
+
+	if filteredZones.Len() < 1 {
+		return fmt.Errorf("there are no valid zones for topology")
+	}
+
+	// Make sure the necessary fields exist
+	if pv.Spec.NodeAffinity == nil {
+		pv.Spec.NodeAffinity = new(v1.VolumeNodeAffinity)
+	}
+
+	if pv.Spec.NodeAffinity.Required == nil {
+		pv.Spec.NodeAffinity.Required = new(v1.NodeSelector)
+	}
+
+	if len(pv.Spec.NodeAffinity.Required.NodeSelectorTerms) == 0 {
+		pv.Spec.NodeAffinity.Required.NodeSelectorTerms = make([]v1.NodeSelectorTerm, 1)
+	}
+
+	// Don't overwrite if topology is already set
+	if len(getTopology(pv, key)) > 0 {
+		return nil
+	}
+
+	pv.Spec.NodeAffinity.Required.NodeSelectorTerms[0].MatchExpressions = append(pv.Spec.NodeAffinity.Required.NodeSelectorTerms[0].MatchExpressions,
+		v1.NodeSelectorRequirement{
+			Key:      key,
+			Operator: v1.NodeSelectorOpIn,
+			Values:   filteredZones.List(),
+		})
+
+	return nil
+}
+
+// translateTopology converts existing zone labels or in-tree topology to CSI topology.
+// In-tree topology has precedence over zone labels.
+func translateTopology(pv *v1.PersistentVolume, topologyKey string) error {
+	zones := getTopology(pv, v1.LabelZoneFailureDomain)
```

Comment:
Need to double check with @andyzhangx , but I recall that azure topology actually needs both zone + region labels to uniquely identify a topology.  It may be that the azure translation cannot reuse this method.

Reply:
Azure CSI drivers didn't support topology back when I started working on PR, so I didn't include Azure PV translation.

Reply:
Apparently it's still not supported in both drivers:

Azure Disk:
https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/master/pkg/azuredisk/nodeserver.go#L306-L309

Azure File:
https://github.com/kubernetes-sigs/azurefile-csi-driver/blob/master/pkg/azurefile/nodeserver.go#L242-L244

Reply:
Can you open an issue on the azure drivers to add topology support and then add the necessary translation logic in k/k?